File: packages/sdk-py/pylva/pylva.py
"""Explicit-client Pylva constructor.

Mirrors packages/sdk-ts/src/Pylva.ts. Telemetry-only deployments
keep using ``pylva.init(api_key)``; cross-provider failover
requires this constructor so the SDK has a handle to the backup
provider's client.

Usage::

    from openai import OpenAI
    from anthropic import Anthropic
    from pylva import Pylva

    Pylva(
        api_key="pv_live_...",
        openai=OpenAI(),
        anthropic=Anthropic(),
        providers={"openrouter": openrouter_client},
    )
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from .core.client_registry import register_provider_client, register_provider_clients
from .core.config import ControlConfig
from .core.control_schema import (
    BudgetCommitRequest,
    BudgetCommitResponse,
    BudgetExtendRequest,
    BudgetExtendResponse,
    BudgetReleaseRequest,
    BudgetReleaseResponse,
    BudgetReservationRequest,
    BudgetReservationResponse,
)


class Pylva:
    """Explicit-client SDK constructor (O6 — required for failover)."""

    def __init__(
        self,
        api_key: str,
        *,
        endpoint: str | None = None,
        batch_size: int = 100,
        flush_interval: float = 5.0,
        local_mode: bool = False,
        non_llm: dict[str, Any] | None = None,
        control: ControlConfig | Mapping[str, Any] | None = None,
        openai: Any | None = None,
        anthropic: Any | None = None,
        providers: Mapping[str, Any] | None = None,
    ) -> None:
        # Late import to avoid circulars; pylva/__init__.py imports
        # this module to re-export Pylva.
        from . import init as _init

        _init(
            api_key,
            endpoint=endpoint,
            batch_size=batch_size,
            flush_interval=flush_interval,
            local_mode=local_mode,
            non_llm=non_llm,
            control=control,
        )

        if openai is not None:
            register_provider_client("openai", openai)
        if anthropic is not None:
            register_provider_client("anthropic", anthropic)
        if providers is not None:
            register_provider_clients(providers)

        self.has_openai = openai is not None
        self.has_anthropic = anthropic is not None

    async def ready(self) -> bool:
        from .core.control_client import ready

        return await ready()

    def ready_sync(self) -> bool:
        from .core.control_client import ready_sync

        return ready_sync()

    async def reserve_usage(
        self,
        request: BudgetReservationRequest | Mapping[str, Any],
    ) -> BudgetReservationResponse:
        from .core.control_client import reserve_usage

        return await reserve_usage(request)

    def reserve_usage_sync(
        self,
        request: BudgetReservationRequest | Mapping[str, Any],
    ) -> BudgetReservationResponse:
        from .core.control_client import reserve_usage_sync

        return reserve_usage_sync(request)

    async def commit_usage(
        self,
        reservation_id: str,
        request: BudgetCommitRequest | Mapping[str, Any],
    ) -> BudgetCommitResponse:
        from .core.control_client import commit_usage

        return await commit_usage(reservation_id, request)

    def commit_usage_sync(
        self,
        reservation_id: str,
        request: BudgetCommitRequest | Mapping[str, Any],
    ) -> BudgetCommitResponse:
        from .core.control_client import commit_usage_sync

        return commit_usage_sync(reservation_id, request)

    async def release_usage(
        self,
        reservation_id: str,
        request: BudgetReleaseRequest | Mapping[str, Any],
    ) -> BudgetReleaseResponse:
        from .core.control_client import release_usage

        return await release_usage(reservation_id, request)

    def release_usage_sync(
        self,
        reservation_id: str,
        request: BudgetReleaseRequest | Mapping[str, Any],
    ) -> BudgetReleaseResponse:
        from .core.control_client import release_usage_sync

        return release_usage_sync(reservation_id, request)

    async def extend_usage(
        self,
        reservation_id: str,
        request: BudgetExtendRequest | Mapping[str, Any],
    ) -> BudgetExtendResponse:
        from .core.control_client import extend_usage

        return await extend_usage(reservation_id, request)

    def extend_usage_sync(
        self,
        reservation_id: str,
        request: BudgetExtendRequest | Mapping[str, Any],
    ) -> BudgetExtendResponse:
        from .core.control_client import extend_usage_sync

        return extend_usage_sync(reservation_id, request)

File: packages/sdk-ts/src/Pylva.ts
// Track 3 PR 3.3 — explicit-client SDK constructor (O6).
//
// New public API:
//
//   import OpenAI from 'openai';
//   import Anthropic from '@anthropic-ai/sdk';
//   import { Pylva } from '@pylva/sdk';
//
//   const pylva = new Pylva({
//     apiKey: 'pv_live_...',
//     openai: new OpenAI(),
//     anthropic: new Anthropic(),
//     providers: { openrouter: openrouterClient },
//   });
//
// The constructor:
//  1. Calls the existing init() so config / accumulator / rules-cache
//     bootstrapping is unchanged.
//  2. Registers the supplied clients in the process-wide registry so the
//     failover engine can pick the correct backup client at call time
//     (D52 / O6). Without explicit registration, cross-provider failover
//     surfaces a warning and falls through to the original error.
//
// Telemetry-only deployments can keep using init({ apiKey }); deployments
// that use reliability_failover rules should use this explicit-client
// constructor so backup-provider clients can be registered.

import { type InitConfig } from './core/config.js';
import { installSdkConfig } from './core/identity.js';
import { applyAllPatches } from './wrappers/_patch.js';
import { initAccumulator } from './core/budget_accumulator.js';
import { ensurePricingCache } from './core/pricing_cache.js';
import { ensureRulesCache, getCachedRules } from './core/rules_cache.js';
import { validateFailoverWrappers } from './wrappers/_init_validation.js';
import { registerProviderClient } from './core/client_registry.js';
import {
  ready,
  controlStatus,
  reserveUsage,
  commitUsage,
  releaseUsage,
  extendUsage,
  type CommitUsageInput,
  type CommitUsageResult,
  type ControlReadyResult,
  type ExtendUsageInput,
  type ExtendUsageResult,
  type ReleaseUsageInput,
  type ReleaseUsageResult,
  type ReserveUsageInput,
  type ReserveUsageResult,
} from './core/control_client.js';

export interface PylvaOptions extends InitConfig {
  /** Builder-instantiated OpenAI client used for failover-targeted calls. */
  openai?: unknown;
  /** Builder-instantiated Anthropic client used for failover-targeted calls. */
  anthropic?: unknown;
  /** Builder-instantiated clients keyed by exact provider id. */
  providers?: Record<string, unknown>;
}

export class Pylva {
  readonly hasOpenAi: boolean;
  readonly hasAnthropic: boolean;

  constructor(options: PylvaOptions) {
    const { openai, anthropic, providers, ...initConfig } = options;
    // Materialize arbitrary provider objects before publishing a new SDK
    // identity. Object.entries can execute Proxy traps or property getters;
    // registration from this detached list is total after installation.
    const providerEntries = providers
      ? Object.entries(providers).map(([provider, client]) => ({ provider, client }))
      : [];
    installSdkConfig(initConfig);
    try {
      applyAllPatches();
    } catch {
      /* R1 — never surface patch errors to host */
    }
    if (openai) registerProviderClient('openai', openai);
    if (anthropic) registerProviderClient('anthropic', anthropic);
    for (let index = 0; index < providerEntries.length; index += 1) {
      const entry = providerEntries[index];
      if (entry) registerProviderClient(entry.provider, entry.client);
    }
    this.hasOpenAi = openai != null;
    this.hasAnthropic = anthropic != null;

    void initAccumulator().catch(() => {
      /* R1 */
    });
    // Warm the pricing cache so local budget accounting (recordLlmSpend)
    // can price calls from the first flush onward.
    void ensurePricingCache().catch(() => {
      /* R1 */
    });
    void ensureRulesCache()
      .then(() => validateFailoverWrappers(getCachedRules()))
      .catch(() => {
        /* R1 */
      });
  }

  ready(): Promise<boolean> {
    return ready();
  }

  controlStatus(): Promise<ControlReadyResult> {
    return controlStatus();
  }

  reserveUsage(input: ReserveUsageInput): Promise<ReserveUsageResult> {
    return reserveUsage(input);
  }

  commitUsage(input: CommitUsageInput): Promise<CommitUsageResult> {
    return commitUsage(input);
  }

  releaseUsage(input: ReleaseUsageInput): Promise<ReleaseUsageResult> {
    return releaseUsage(input);
  }

  extendUsage(input: ExtendUsageInput): Promise<ExtendUsageResult> {
    return extendUsage(input);
  }
}

Object.defineProperty(Pylva, 'name', { value: 'Pylva' });
